### PR TITLE
document unavailability of mount propogation

### PR DESCRIPTION
**Pull request by doehyunbaek**
Hi. I really did write small writeup concerning unavailability of mount propagation on Docker Desktop with a pointer to the issue thread(https://github.com/kubernetes-sigs/kind/issues/2576).

The thing I'm confused is, is the note below still relevant? As I read your explaniation on the issue thread, setting hostPath in Docker desktop configuration would not really help. Is there something I am missing? If that explaination is misleading, maybe I should delete that note as well? 

https://github.com/kubernetes-sigs/kind/blob/80836a7ecbadcc3864ae295ff8e1a06877e81674/site/content/docs/user/configuration.md?plain=1#L287-L290

### CODE REVIEW

**Comment by doehyunbaek** [comment]
closes #2576

**Comment by BenTheElder** [comment]
> The thing I'm confused is, is the note below still relevant? As I read your explaniation on the issue thread, setting hostPath in Docker desktop configuration would not really help. Is there something I am missing? If that explaination is misleading, maybe I should delete that note as well?

Yes it is still relevant, you cannot mount a path from the host that is not shared in the docker desktop configuration.
That's not related to mount propagation.

File: site/content/docs/user/configuration.md [hunk 1]
```diff
@@ -289,6 +289,11 @@ included in the Preferences -> Resources -> File Sharing.
 
 For more information see the [Docker file sharing guide.](https://docs.docker.com/docker-for-mac/#file-sharing)
 
+**NOTE**: Mount propagation is not possible with Mac or Windows Docker Desktop,
+due to limitation of Docker Desktop implementation.
```
Comment (BenTheElder) [inline site/content/docs/user/configuration.md:293]: I'm not sure how to tersely cover this off the top of my head but, this is not quite right.

You cannot use mount propagation *when the mount is to a path on the mac or windows host* *when using docker desktop* (and presumably other tools, but I can't say for certain, I don't other VM <> host boundaries can handle this either).

When using docker desktop you could still use this setting *if* the hostPath is to a location on the VM and not the mac/windows host. Which is generally unusual I think, but possible.

If #1966 lands, it should also be possible for that, because volumes are generally contained within the VM disk.

Comment (BenTheElder) [reply site/content/docs/user/configuration.md:293]: So basically ... mount propagation doesn't work across the macOS / Windows | Docker Desktop Virtual Machine boundary. Mounts that cross that boundary can't be expected to support this, and that limitation comes from Docker.

**Comment by doehyunbaek** [comment]
Thanks for your quick and kind reply. Here are some thoughts,

> When using docker desktop you could still use this setting if the hostPath is to a location on the VM and not the mac/windows host. Which is generally unusual I think, but possible.

I am unaware of this possibility. If you know any way to refer to VM path in docker cli with Docker desktop for mac(God this is wordy XD). Can you let me know? I have tinkered with both -v and --mount option and could not find success. From design perspective also it seems unreasonable for docker folks to let source be physical(mac) machine path and virtual(linux) machine path. Maybe this is the reason why docker docs resorts to just 'Mount propagation doesn't work with docker desktop'(https://github.com/docker/docker.github.io/pull/14355). And I think this statement is not only useful but also valid if there is no way to specificy vm path as bind source.

I think it is true that mount propagation is pretty advanced configuration as mentioned in [docker docs](https://docs.docker.com/storage/bind-mounts/#:~:text=Bind%20propagation%20is%20an%20advanced%20topic%20and%20many%20users%20never%20need%20to%20configure%20it.) and having this setting as HostToContainer [example](https://github.com/kubernetes-sigs/kind/blob/47d47043e041536ffa5d64730ff593e991130de3/site/static/examples/config-with-mounts.yaml#L18) trips many people up. So my suggestion is setting this option in example to None, as that would prevent many newcomers copy-pasting example and finding out it does not work and getting into rabbit hole world of linux mount propagation.

Also the fact that bind propagation would be mostly useless for non-advanced usecases(I tried to find compelling example for its use but couldn't find any... one I got [here](https://medium.com/kokster/kubernetes-mount-propagation-5306c36a4a2d#:~:text=available%20here.-,Use%20Cases,-Mount%20propagation%20mode) seems to be not so relevant for beginners.) supports this approach. And I presume people who need these advanced usecases would be familiar with these topics and could advance in their own with a little pointer to these threads.

So In summary, I suggest

1. Chaning propagation option in config-with-mounts.yaml from HostToContainer to None.

2-a. Provided there is no way to configure vm path as bind source, just say 'Bind propagation is not possible for docker desktop for mac or windows'

2-b. If there is some arcane way to configure vm path as bind source, be more specific and say 'Bind propagation does not work across VM boundaries. e.g. You can not bind propagate your mac host directory into container. use None option instead'.

Again, thanks for your past and current reply. Learned a lot investigating :)

**Comment by BenTheElder** [comment]
> I am unaware of this possibility. If you know any way to refer to VM path in docker cli with Docker desktop for mac(God this is wordy XD), can you let me know?

KIND has some internally, for example `--volume /lib/modules:/lib/modules` or `--volume /dev/mapper:/dev/mapper` when appropriate. `/lib/modules` is the Linux VM's kernel modules and `/dev/mapper` the VM's device mapper, not something from the host (these don't exist on macOS or Windows).

> From design perspective also it seems unreasonable for docker folks to let source be both physical(mac) machine path and virtual(linux) machine path.

Not at all ... If the path isn't shared from mac and does exist on the VM, it makes sense to mount it from the VM, this just makes it behave the same as docker on linux without the docker desktop magic. The docker desktop filesystem magic comes into play for paths configured in the application.


> I think it is true that mount propagation is pretty advanced configuration as mentioned in [docker docs](https://docs.docker.com/storage/bind-mounts/#:~:text=Bind%20propagation%20is%20an%20advanced%20topic%20and%20many%20users%20never%20need%20to%20configure%20it.) and having this setting as HostToContainer [example](https://github.com/kubernetes-sigs/kind/blob/47d47043e041536ffa5d64730ff593e991130de3/site/static/examples/config-with-mounts.yaml#L18) trips many people up. So my suggestion is setting this option in example to None, as that would prevent many newcomers copy-pasting example and finding out it does not work and getting into rabbit hole world of linux mount propagation.


Sure, setting the default to `None` would be fine.

I think better yet, we should have an example of reasonable *actual* usage that doesn't bother with unnecessary fields, however we still need to document that they're available.

**Comment by doehyunbaek** [comment]
Ah, my confusion was that I thought source path in -v option refers to path in physical machine, but apparently it is not. As can be shown from below example,
```
# fails because sharing is not configured and magic does not work.
$ docker run -it -v /Applications:/don ubuntu bash
docker: Error response from daemon: Mounts denied: 
The path /Applications is not shared from the host and is not known to Docker.
You can configure shared paths from Docker -> Preferences... -> Resources -> File Sharing.
See https://docs.docker.com/desktop/mac for more info.
ERRO[0000] error waiting for container: context canceled 
# succeeds because sharing is configured for /Users 
$ docker run -it -v /Users:/don ubuntu bash
/ # 
# succeeds because source path refers to /lib/modules in vm, not mac physical machine. /lib/modules does not exist in physical machine.
bash-5.1$ docker run -it -v /lib/modules:/don ubuntu bash
root@ad1d52570788:/# ls /don
5.10.104-linuxkit
```

It seems to make more sense this way as mount is happening at where the docker daemon is hosted not where cli lives. Thanks for correction.

I'm not sure how to progress on this issue though. As of actual usage, I am pretty fine using kind cluster with not much confiuration in my daily usage, so I wouldn't be able to find one in my configuration.  But I can look for any sensible usage of kind.x-k8s.io/v1alpha4 configuration throught github.

**Comment by doehyunbaek** [comment]
I'm closing this PR as I'm not sure how to progress. Will reopen if I get some clue.
